### PR TITLE
Create a script to convert from nnUNet V2 to BIDS

### PR DESCRIPTION
**Pull request by tzebre**

### CODE REVIEW

**Comment by tzebre** [comment]
In the case of a region based segmentation training the `dataset.json` look like this 
```
...
"labels": {
    "background": 0,
    "whole_tumor": [1, 2, 3],
    "tumor_core": [2, 3],
    "enhancing_tumor": 3  # or [3]
},
"regions_class_order": [1, 2, 3],
...
```

So the label file corresponding to the image has 3 voxel values (1,2,3). 

My idea is to separate this file in 4 file:
- `...whole_tumor_seg.nii.gz` with voxel >=1
- `...tumor_core_seg.nii.gz` with voxel >=2
- `...enhancing_tumor_seg.nii.gz`with voxel =3
- `...multi_seg.nii.gz`the orginal file from nnUNet dataset

**Comment by jcohenadad** [comment]
I disagree with https://github.com/ivadomed/data-conversion/pull/15#issuecomment-1599351103, because this example is specific to a segmentation case (whole_tumor, enhancing_tumor, etc.). What if we would like to use this conversion script with single class MS lesion seg? Or 2-class white/gray matter? etc.

**Comment by tzebre** [comment]
> I disagree with [#15 (comment)](https://github.com/ivadomed/data-conversion/pull/15#issuecomment-1599351103), because this example is specific to a segmentation case (whole_tumor, enhancing_tumor, etc.). What if we would like to use this conversion script with single class MS lesion seg? Or 2-class white/gray matter? etc. The code should be made more generic.

I am not sure to understand the problem. The conversion process depend on the `dataset.json` file, so no matter the number of classes the result is based on the informations provided in the `dataset.json`. 

For example if your dataset has 2 class segmentation :
- `dataset.json`
```
"labels": {
    "background": 0,
    "white_matter": 1,
    "grey_matter":2
}
```
- Created label file in the Bids dataset : 
 - `...white_matter.nii.gz` with voxel = 1
 - `...grey_matter.nii.gz` with voxel = 2
 - `...multi_seg.nii.gz` the original file

**Comment by jcohenadad** [comment]
> > I disagree with [#15 (comment)](https://github.com/ivadomed/data-conversion/pull/15#issuecomment-1599351103), because this example is specific to a segmentation case (whole_tumor, enhancing_tumor, etc.). What if we would like to use this conversion script with single class MS lesion seg? Or 2-class white/gray matter? etc. The code should be made more generic.
> 
> I am not sure to understand the problem. The conversion process depend on the `dataset.json` file, so no matter the number of classes the result is based on the informations provided in the `dataset.json`.
> 
> For example if your dataset has 2 class segmentation :
> 
> * `dataset.json`
> 
> ```
> "labels": {
>     "background": 0,
>     "white_matter": 1,
>     "grey_matter":2
> }
> ```
> 
> * Created label file in the Bids dataset :
> * `...white_matter.nii.gz` with voxel = 1
> * `...grey_matter.nii.gz` with voxel = 2
> * `...multi_seg.nii.gz` the original file

Nevermind, I thought your code was actually including these hard-coded classes, but you just gave an example of a JSON file to use with your code-- I suggest you always write a quick intro/context/purpose in the [body of your PR](https://github.com/ivadomed/data-conversion/pull/15#issue-1765280315) so reviewers know what the PR is about. Thanks

File: dataset_conversion/nnunetV2_to_bids.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+def main():
+    parser = get_parser()
+    args = parser.parse_args()
+    copy = args.copy
+    suffix =args.suffix
+    root = Path(os.path.abspath(os.path.expanduser(args.path_bids)))
+    path_out = Path(os.path.abspath(os.path.expanduser(args.path_out)))
+    with open(os.path.join(root, "dataset.json"), 'r') as json_file:
+        dataset_info = json.load(json_file)
```
Comment (valosekj) [inline dataset_conversion/nnunetV2_to_bids.py:71]: A check if the `dataset.json` file exists could be added, raising an error if the file does not exist.

File: dataset_conversion/nnunetV2_to_bids.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+import pathlib
+from pathlib import Path
+import json
+import os
+import nibabel as nib
+
+def get_parser():
+    # parse command line arguments
+    parser = argparse.ArgumentParser(description='Convert BIDS-structured dataset to nnUNetV2 database format.')
+    parser.add_argument('--path-bids', required=True, help='Path to nnUNet dataset. Example: ~/data/dataset')
```
Comment (valosekj) [inline dataset_conversion/nnunetV2_to_bids.py:20]:
```suggestion
    parser.add_argument('--path-in', required=True, help='Path to nnUNet dataset. Example: ~/data/dataset')
```

File: dataset_conversion/nnunetV2_to_bids.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import shutil
+import pathlib
+from pathlib import Path
+import json
+import os
+import nibabel as nib
+
+def get_parser():
+    # parse command line arguments
+    parser = argparse.ArgumentParser(description='Convert BIDS-structured dataset to nnUNetV2 database format.')
```
Comment (valosekj) [inline dataset_conversion/nnunetV2_to_bids.py:19]:
```suggestion
    parser = argparse.ArgumentParser(description='Convert dataset in nnUNetV2 format to BIDS-structured dataset.')
```

File: dataset_conversion/nnunetV2_to_bids.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    return parser
+
+
+
+def get_subject_info(file_name, contrast_dict):
+    """
+    Get different information about the current subject
+
+    Args:
+        file_name (str): Filename corresponding to the subject image
```
Comment (valosekj) [inline dataset_conversion/nnunetV2_to_bids.py:36]: Could you add an example file_name here?

File: dataset_conversion/nnunetV2_to_bids.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        contrast_dict (dict): Dictionary, key channel_names from dataset.json
+
+    Returns:
+        sub_names (str): Name of the subject. Example: sub-milan002
+        ses (str): ses name
```
Comment (valosekj) [inline dataset_conversion/nnunetV2_to_bids.py:41]:
```suggestion
        ses (str): session name. Example: ses-01
```

File: dataset_conversion/nnunetV2_to_bids.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        bids_nb (str): subject number in the BIDS dataset
+        info[2] (str): Contrast value in BIDS format. Example 0001
```
Comment (valosekj) [inline dataset_conversion/nnunetV2_to_bids.py:43]:
```suggestion
        info[2] (str): Contrast value in BIDS format. Example: 0001
```

File: dataset_conversion/nnunetV2_to_bids.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        contrast = '0'
+    contrast_bids = contrast_dict[contrast]
+    return sub_name, ses, bids_nb, contrast, contrast_bids
+
+
+def main():
+    parser = get_parser()
+    args = parser.parse_args()
+    copy = args.copy
+    suffix =args.suffix
```
Comment (valosekj) [inline dataset_conversion/nnunetV2_to_bids.py:67]:
```suggestion
    suffix = args.suffix
```

File: dataset_conversion/nnunetV2_to_bids.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
+"""
+Converts nnUNetv2 dataset format to the BIDS-structured dataset. Full details about
+the format can be found here: https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/dataset_format.md
+
```
Comment (valosekj) [inline dataset_conversion/nnunetV2_to_bids.py:4]: We can add an example

```suggestion

nnUNetV2 format:

├── dataset.json
├── imagesTr
│   ├── sub-001_ses-01_001_0000.nii.gz
...
├── imagesTs
│   ├── sub-004_ses-01_001_0000.nii.gz
...
├── labelsTr
│   ├── sub-001_ses-01_001.nii.gz
...
├── labelsTs
    ├──sub-zh004_ses-01_001.nii.gz
...

BIDS format:

├── derivatives
│   └── labels
│       ├── sub-001
│       │   ├── ses-01
│       │   │   └── anat
│       │   │       └── sub-001_ses-01_T2w_seg.nii.gz
...
├── sub-001
│   ├── ses-01
│   │   └── anat
│   │       └── sub-001_ses-01_T2w.nii.gz
...

```

**Review by valosekj — APPROVED** [review]
Thank you @tzebre! LGTM! I just left a few minor comments.

File: dataset_conversion/nnunetV2_to_bids.py (hunk 4 above)
Comment (valosekj) [reply dataset_conversion/nnunetV2_to_bids.py]: Done in https://github.com/ivadomed/utilities/pull/15/commits/5d0b9fcbb66554eb32982e42582eef28fa676e6b

File: dataset_conversion/nnunetV2_to_bids.py (hunk 1 above)
Comment (valosekj) [reply dataset_conversion/nnunetV2_to_bids.py:104]: Done in https://github.com/ivadomed/utilities/pull/15/commits/5d0b9fcbb66554eb32982e42582eef28fa676e6b

File: dataset_conversion/nnunetV2_to_bids.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
```
Comment (valosekj) [inline dataset_conversion/nnunetV2_to_bids.py:4]:
```suggestion
This script converts the nnUNetv2 dataset format to the BIDS-structured dataset. 
Details about the nnUNetv2 format can be found here: https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/dataset_format.md
Details about the BIDS standard can be found here: https://bids.neuroimaging.io

```

**Pull request merged**